### DISCUSSION TITLE
[LAMMPS] Benchmarking Same As NVIDIA

### DISCUSSION THREAD

**Discussion by AnuragKr** [deepmodeling/deepmd-kit#2007] [Q&A] [answered]
Hello,

In the [link](https://developer.nvidia.com/hpc-application-performance) provided, NVIDIA has benchmarked various MD simulation software. I kind of want to replicate the result in the same way with DeePMD-kit on V100. So, first, I attempted to replicate the same result that they had.

LAMMPS Code --
```lammps
# 3d Lennard-Jones melt

variable        x index 1
variable        y index 1
variable        z index 1

variable        xx equal 20*$x
variable        yy equal 20*$y
variable        zz equal 20*$z

units           lj
atom_style      atomic

lattice         fcc 0.8442
region          box block 0 ${xx} 0 ${yy} 0 ${zz}
create_box      1 box
create_atoms    1 box
mass            1 1.0

velocity        all create 1.44 87287 loop geom

pair_style      lj/cut 2.5
pair_coeff      1 1 1.0 1.0 2.5

neighbor        0.3 bin
neigh_modify    delay 0 every 20 check no

fix             1 all nve

run             100
```
Command to run Lammps

```bash
mpirun -np ${gpu_count} lmp -k on g ${gpu_count} -sf kk -pk kokkos cuda/aware on neigh full comm device binsize 2.8 -var x 8 -var y 4 -var z 8 -in ${input}
```
But they are using kokkos package to produce result. Is it possible to produce same result as kokkos package in ATOM-Time Steps/s as a metric by running Lammps normally like ---

```bash
mpiexec -np ${gpu_count} lmp -in ${input}
```
What changes are required to achieve the above-mentioned result? How do I add the kokkos package to an existing conda DeePMD-kit LAMMPS installation? Is it possible to add additional packages only if I install LAMMPS as a plugin?

**Comment by njzjz** [comment, marked as answer]
Why do you want to use kokkos?

It looks like kokkos does not provide a plugin, so if you want to use kokkos you have to recompile LAMMPS.

**Reply by AnuragKr** [reply]
Q.) Why do you want to use kokkos?
  -- I just wanted to run exactly how NVIDIA ran.
Q.) It looks like kokkos does not provide a plugin, so if you want to use kokkos you have to recompile LAMMPS.
 -- How to do that if I have done installation via conda ?

One doubt -- when I run lammps via mpirun -np 2 lmp -in in.lmp  Does this command implicitly use GPU although from nvidia-smi I can see GPU is getting utilized. But then why Lammps has given different command if we want to run using GPU.

And we don't need kokkos as simple running command can also serve our purpose. Or is it something kokkos run it efficiently.

**Reply by njzjz** [reply]
> when I run lammps via mpirun -np 2 lmp -in in.lmp Does this command implicitly use GPU although from nvidia-smi I can see GPU is getting utilized. 

In this case only deepmd-kit utilize GPUs. LAMMPS function still uses CPUs.

**Reply by AnuragKr** [reply]
I ran following LAMMPS code using DeepMD-kit and none of the GPU getting utilized --
```lammps
# 3d Lennard-Jones melt

variable        x index 1
variable        y index 1
variable        z index 1

variable        xx equal 20*24
variable        yy equal 20*12
variable        zz equal 20*16

units           lj
atom_style      atomic

lattice         fcc 0.8442
region          box block 0 ${xx} 0 ${yy} 0 ${zz}
create_box      1 box
create_atoms    1 box
mass            1 1.0

velocity        all create 1.44 87287 loop geom

pair_style      lj/cut 2.5
pair_coeff      1 1 1.0 1.0 2.5

neighbor        0.3 bin
neigh_modify    delay 0 every 20 check no

fix             1 all nve

run             10000
```
Command I ran DeepMD conda environment-- 
```bash
mpiexec -np 8 lmp -in in.lmp
```
GPU Statistics

![Screenshot 2022-10-20 at 11 21 05 AM](https://user-images.githubusercontent.com/10074710/196866743-5571694d-f325-449e-b220-cb4d8db63dd5.png)

Is it the case GPU will be utilized only when we will provide DeePMD potential in pair style argument than any other potential. Please let me know if I am missing out on something.

**Reply by njzjz** [reply]
> Is it the case GPU will be utilized only when we will provide DeePMD potential in pair style argument than any other potential.

Yes

**Reply by AnuragKr** [reply]
@njzjz  Is it possible to enable GPU support for default potentials provided for LAMMPS inside DeePMD? If yes what sort of changes do we need to do?

**Reply by njzjz** [reply]
We do not intend to do it so you have to compile lammps by yourself.

**Reply by AnuragKr** [reply]
1. Could you please tell me why default potential support was not added?
2.  I intend to make generalize the LAMMPS plugin so that it can run both, Is it possible?

**Reply by njzjz** [reply]
1. We are not intended to distribute something out of the DeePMD-kit field.
2. Yes, but note that LAMMPS plugin needs the same compiler, C++ standard, and MPI standard.

**Reply by AnuragKr** [reply]
Could you please tell me how the LAMMPS plugin was developed in the DeepMD-kit-only binary file I can see deepmd_lmp where the code of that section is written.

**Reply by njzjz** [reply]
In `source/lmp`
